Updated EiffelBase classes `ISE_RUNTIME`, `REFLECTOR` and `REFLECTED_…

jvelilla opened the pull request:
…COPY_SEMANTICS_OBJECT`

which are implemented with external C to use built-in routines.

jvelilla committed:
Updated EiffelBase classes `ISE_RUNTIME`, `REFLECTOR` and `REFLECTED_COPY_SEMANTICS_OBJECT`
which are implemented with external C to use built-in routines.

jocelyn commented on Src/Delivery/studio/built_ins/classic/ISE_RUNTIME.e:
```diff
@@ -0,0 +1,81 @@
+		external
+			"C inline use %"eif_out.h%""
+		alias
+			"return c_generator_of_type(eif_decoded_type($a_type_id));"
```
Is there any need for the instance_free postcondition ?
or this is really part of the original class?

jvelilla replied:
Good question, I've checked other built-ins and I didn't see they use the instance free postcondition,  the original class has the instance free postcondition.
Similarly, some features are marked as frozen in the original class and I didn`t use it in the built-in class.

jocelyn replied:
I would say ... if the compiler accepts the frozen and the assertion in the built-in classes, then add them. At worst they will be ignored.
But be sure to keep them in the original Eiffel class.

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -193,8 +202,23 @@ rt_private rt_inline EIF_BOOLEAN rt_is_special_copy_semantics_item (EIF_INTEGER_
-#define eif_builtin_ISE_RUNTIME_set_pointer_field_at(field_offs,obj,offs,val)			*(EIF_POINTER *) (eif_obj_at(obj,offs) + field_offs) = (EIF_POINTER) (val)
+#define eif_builtin_ISE_RUNTIME_set_pointer_field_at(field_offs,obj,offs,val)		*(EIF_POINTER *) (eif_obj_at(obj,offs) + field_offs) = (EIF_POINTER) (val)
 
+#define eif_builtin_ISE_RUNTIME_unlock_marking							eif_unlock_marking
```
it would have help for the review, to have logic order in this file.
Either same order as original ISE_RUNTIME,
or alphabetical order.

jvelilla replied:
I just add the new built-in features at the end of the  `ISE_RUNTIME` section, but if it`s clearer to have them in alphabetical order I can do it.

jocelyn commented on Src/library/base/elks/kernel/ise_runtime.e:
```diff
@@ -812,9 +762,7 @@ feature -- Object marking
-			"C blocking use %"eif_traverse.h%""
-		alias
-			"eif_lock_marking"
+			"built_in static"
```
How do you handle the "blocking" semantic?

jvelilla replied:
Good question, I will move this code to the built-in class.

jvelilla replied:
Done

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -193,8 +202,23 @@ rt_private rt_inline EIF_BOOLEAN rt_is_special_copy_semantics_item (EIF_INTEGER_
+#define eif_builtin_ISE_RUNTIME_set_pointer_field_at(field_offs,obj,offs,val)		*(EIF_POINTER *) (eif_obj_at(obj,offs) + field_offs) = (EIF_POINTER) (val)
 
+#define eif_builtin_ISE_RUNTIME_unlock_marking							eif_unlock_marking
+#define eif_builtin_ISE_RUNTIME_lock_marking							eif_lock_marking
```
the original code in ISE_RUNTIME is using  ` 			"C blocking use %"eif_traverse.h%""   `
so maybe this one should be move to the ISE_RUNTIME.e builtin file instead.

jvelilla replied:
Yes, initially I move it to `eif_built_in.h`, but I don`t know how to handle `blocking` in the macro so maybe the best option is to move it to `ISE_RUNTIME.e` builtin file.

jvelilla replied:
Done

jocelyn commented on Src/Delivery/studio/built_ins/classic/ISE_RUNTIME.e:
```diff
@@ -0,0 +1,81 @@
+			"c_check_assert"
+		end
+
+	in_assertion: BOOLEAN
```
Why did you kept this in_assertion here, and not moved to eif_built_in.h ?
(just wondering)

jvelilla replied:
Just because it has two instructions, and I thought it was easier to read in the ISE_RUNTIME.e file than in the 
eif_built_inh.
```
#define eif_builtin_ISE_RUNTIME_in_assertion{\
        GTCX;\ /* --comment */
        return EIF_TEST(in_assertion!=0);\
    }
```

jocelyn replied:
ok, good reason. I could live with both solutions.

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -193,8 +202,23 @@ rt_private rt_inline EIF_BOOLEAN rt_is_special_copy_semantics_item (EIF_INTEGER_
+#define eif_builtin_ISE_RUNTIME_generic_parameter_count(a_type_id)		eif_gen_count_with_dftype(eif_decoded_type(a_type_id).id)
+#define eif_builtin_ISE_RUNTIME_field_offset_of_type(i,a_type_id)		ei_offset_of_type(i,a_type_id)
+#define eif_builtin_ISE_RUNTIME_field_count_of_type(a_type_id)			ei_count_field_of_type(a_type_id)
+#define eif_builtin_ISE_RUNTIME_compiler_version						egc_compiler_tag
```
original code was
		external
			"C macro use %"eif_project.h%""
		alias
			"egc_compiler_tag"


is eif_built_in.h include the eif_project.h ?

jvelilla replied:
I will move it to the builtin class since eif_project.h is not part of  eif_built_in.h

jvelilla replied:
Done

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -193,8 +202,23 @@ rt_private rt_inline EIF_BOOLEAN rt_is_special_copy_semantics_item (EIF_INTEGER_
+#define eif_builtin_ISE_RUNTIME_persistent_field_count_of_type(a_type_id)	(System(To_dtype(eif_decoded_type(a_type_id).id)).cn_persistent_nbattr)
+#define eif_builtin_ISE_RUNTIME_is_field_expanded_of_type(i,a_type_id)		((System(To_dtype(eif_decoded_type(a_type_id).id)).cn_types[i - 1] & SK_HEAD) == SK_EXP)
+#define eif_builtin_ISE_RUNTIME_is_field_transient_of_type(i,a_type_id)		EIF_IS_TRANSIENT_ATTRIBUTE(System(To_dtype(eif_decoded_type(a_type_id).id)), i - 1)
+#define eif_builtin_ISE_RUNTIME_attached_type(a_type_id)					eif_attached_type(a_type_id)
```
please align the definition ... (bad indentation)

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -193,8 +202,23 @@ rt_private rt_inline EIF_BOOLEAN rt_is_special_copy_semantics_item (EIF_INTEGER_
-#define eif_builtin_ISE_RUNTIME_set_pointer_field_at(field_offs,obj,offs,val)			*(EIF_POINTER *) (eif_obj_at(obj,offs) + field_offs) = (EIF_POINTER) (val)
+#define eif_builtin_ISE_RUNTIME_set_pointer_field_at(field_offs,obj,offs,val)		*(EIF_POINTER *) (eif_obj_at(obj,offs) + field_offs) = (EIF_POINTER) (val)
 
+#define eif_builtin_ISE_RUNTIME_unlock_marking							eif_unlock_marking
```
I suggest that you update the order, eventually move define up or down ... to have logical order.

jocelyn commented on Src/library/base/elks/support/reflected_copy_semantics_object.e:
```diff
@@ -118,20 +118,16 @@ feature {NONE} -- Implementation
-			"C inline use %"eif_built_in.h%""
-		alias
-			"return *(EIF_REFERENCE *) eif_obj_at($ptr, $offset);"
+			"built_in static"
```
I thought it was eventually possible to reuse the {ISE_RUNTIME}.raw_reference_field_at_offset ?

jvelilla replied:
I will check

jvelilla replied:
ok here we can do
```
dereference (ptr: POINTER; offset: INTEGER): POINTER
		do
			Result :=  {ISE_RUNTIME}.raw_reference_field_at_offset (ptr, offset)
		end
```

jvelilla replied:
Done

jocelyn replied:
I think the more we have in the Eiffel classes, the better it is.
built-in facility is really useful for portability, and for compatibility with Gobo,
but when using it .. it is not nice to see what is really implemented.

jocelyn replied:
maybe we could improve EiffelStudio to show the real implementation.

jocelyn commented on Src/library/base/elks/support/reflected_copy_semantics_object.e:
```diff
@@ -118,20 +118,16 @@ feature {NONE} -- Implementation
-			"C inline use %"eif_built_in.h%""
-		alias
-			"return (EIF_REFERENCE) $ptr;"
+			"built_in static"
```
I thought it was eventually possible to reuse the {ISE_RUNTIME}.reference_field_at_offset ?

jvelilla replied:
I will check

jvelilla replied:
so `return (EIF_REFERENCE) $ptr;"` is equal to `return *(EIF_REFERENCE *) eif_obj_at($ptr,0));` if yes I can do  `Result :=  {ISE_RUNTIME}.raw_reference_field_at_offset (ptr, 0)`

jocelyn commented on Src/C/run-time/eif_built_in.h:
```diff
@@ -88,7 +88,16 @@ extern "C" {
+#define eif_builtin_REFLECTOR_c_new_tuple_instance_of(type_id)			RTLNT(eif_decoded_type(type_id).id)
+#define eif_builtin_REFLECTOR_c_new_instance_of(type_id)			RTLNSMART(eif_decoded_type(type_id).id)
+#define eif_builtin_REFLECTOR_field_count_of_type(type_id)			ei_count_field_of_type(type_id)
+#define eif_builtin_REFLECTOR_field_count_of_type(type_id)			ei_count_field_of_type(type_id)
```
Same line as before ...
I am surprised the C compiler did not report any warning .

jvelilla replied:
I've compiled it and I didn't notice it but .. good catch I will update it.

jvelilla replied:
Done

jvelilla committed:
Updated eif_built_in.h file removed duplicated entry `eif_builtin_REFLECTOR_field_count_of_type(type_id)` and moved built-in features
`compiler_version` and `lock_marking` to the built-in class `ISE_RUNTIME`.
Updated `{REFLECTED_COPY_SEMANTICS_OBJECT}.dereference` feature to use the
`{ISE_RUNTIME}.raw_reference_field_at_offset` feature.
